### PR TITLE
[layer-internal] refactoring weight, input, output to layer_node

### PR DESCRIPTION
**Pull request by lhs8928**
- Implement getter for variable and grad of weight, input, output
- Implement getNumWeights

Self evaluation:

Build test: [X]Passed [ ]Failed [ ]Skipped
Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/models/neuralnet.cpp [hunk 1]
```diff
@@ -577,20 +576,19 @@ int NeuralNetwork::train_run() {
...
+  auto &label = last_layer_node->getOutputGrad(0);
+  auto &in = first_layer_node->getInput(0);
 
   /// @todo migrate this to trait based system; sth like need label?
 
   for (auto iter = model_graph.cbegin(); iter != model_graph.cend(); iter++) {
-    auto const &layer_ = (*iter)->getObject();
-    if (layer_->getType() == "centroid_knn") {
-      layer_->net_hidden[0]->getGradientRef() = label;
+    if ((*iter)->getType() == "centroid_knn") {
```
Comment (kparichay) [inline nntrainer/models/neuralnet.cpp:590]:
```suggestion
    if ((*iter)->getObject()->requireLabel()) {
```

Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:590]: consider for a later PR.

Comment (zhoonit) [reply nntrainer/models/neuralnet.cpp:590]: This is merged, @lhs8928  could you cherry pick https://github.com/nnstreamer/nntrainer/pull/1263 to make the PR complete?

File: nntrainer/layers/layer_node.h [hunk 2]
```diff
@@ -165,6 +165,19 @@ class LayerNode : public ml::train::Layer, public GraphNode {
    */
   unsigned int getNumOutputs() const { return output_layers.size(); }
 
+  /**
+   * @brief Get the number of weights
+   *
+   * @return unsigned int number of weights
+   */
+  unsigned int getNumWeights() const {
+    if (LAYER_V2) {
```
Comment (zhoonit) [inline nntrainer/layers/layer_node.h:174]: Discussed offline to request change this to runtime switch.

sthlike below.
if (layerV1==nullptr) {
}

**Review by zhoonit — COMMENTED** [review]
Great job, waiting for above concerns are resolved :)

File: nntrainer/models/neuralnet.cpp (hunk 1 above)
Comment (lhs8928) [reply nntrainer/models/neuralnet.cpp]: Done

File: nntrainer/layers/layer_node.h (hunk 2 above)
Comment (lhs8928) [reply nntrainer/layers/layer_node.h:174]: Discussed offline that @zhoonit will take this

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**